File: 0662-maximum-width-of-binary-tree/0662-maximum-width-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    using treeIdx = pair<TreeNode*,long long>;
    long long mod = INT_MAX;
    int widthOfBinaryTree(TreeNode* root) {
        // 1. consider each node as row in 2D matrix 
        // 2. left node idx = 2*pidx, right node idx = 2*pidx+1;
        queue<treeIdx> q; 
        q.push({root,0}); 
        long long res = 0,n;
        // TreeNode *top; 
        while(q.size()){
            n=q.size();
            long long fidx = q.front().second;
            for(int i=0;i<n;i++){
                TreeNode* top= q.front().first;
                long long idx = q.front().second;
                res=max(abs(fidx-idx)+1,res);
                q.pop();
                idx = idx-fidx;
                if(top->left) q.push({top->left,(2*idx)});
                if(top->right) q.push({top->right,((2*idx)+1)});
            }
        }
        return res;
    }
};

File: 0662-maximum-width-of-binary-tree/0662-maximum-width-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:

    def dfs(self,node,coord,ridx,nodemap): 
        if node is None: 
            return 
        
        if ridx in nodemap: 
            l,r = nodemap[ridx] 
            nodemap[ridx]=(min(l,coord),max(r,coord))
        else: 
            nodemap[ridx]=(coord,coord)
        
        self.dfs(node.left,2*coord,ridx+1,nodemap) 
        self.dfs(node.right,2*coord+1,ridx+1,nodemap)


    def widthOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        nodemap = {} 
        self.dfs(root,1,1,nodemap)
        res = 0 
        for k,coords in nodemap.items(): 
            l,r=coords 
            res = max(res,r-l+1)
        return res 